packages/validation/src/validateField/conditions/testConditionalRestriction.ts

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2024 The Ontario Institute for Cancer Research. All rights reserved
+ *
+ * This program and the accompanying materials are made available under the terms of
+ * the GNU Affero General Public License v3.0. You should have received a copy of the
+ * GNU Affero General Public License along with this program.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import {
+	ARRAY_TEST_CASE_DEFAULT,
+	ArrayTestCase,
+	type ArrayDataValue,
+	type ConditionalRestrictionTest,
+	type DataRecord,
+	type DataRecordValue,
+	type RestrictionCondition,
+	type SingleDataValue,
+} from '@overture-stack/lectern-dictionary';
+import { resultForArrayTestCase } from '../../utils/resultForArrayTestCase';
+import { testMatchCount } from './testMatchCount';
+import { testMatchCodeList } from './testMatchCodeList';
+import { testMatchExists } from './testMatchExists';
+import { testMatchRange } from './testMatchRange';
+import { testMatchRegex } from './testMatchRegex';
+import { testMatchValue } from './testMatchValue';
+
+/**
+ * Test values extracted from other fields vs a match test. This function should be passed the
+ *
+ * @param values
+ * @param rule
+ * @param matchTest
+ * @param arrayCase  Intentionally using `| undefined` vs `/?: arrayCase` to ensure that this argument is provided when this function is called
+ * @returns
+ */
+const fieldsPassMatchTest = (
+	values: DataRecordValue[],
+	matchTest: (value: DataRecordValue) => boolean,
+	arrayCase: ArrayTestCase | undefined,
+): boolean => {
+	const fieldTestResults = values.map((value) => matchTest(value));
+	return resultForArrayTestCase(fieldTestResults, arrayCase || ARRAY_TEST_CASE_DEFAULT);
+};
+
+const testConditionForSingularValue = (
+	condition: RestrictionCondition,
+	_value: SingleDataValue,
+	fieldValues: DataRecordValue[],
+): boolean => {
```

Comment:
This is the function that tests a condition to see if which restrictions to apply.

The specific code for each of the matching cases are in dedicated files in this same dir.

The `_value` parameter here is the value of the field being tested. It is here in case needed for future use cases such as comparing its value vs another field to determine if restrictions are needed. This is likely to be removed in the next iteration. Since this function is not exported there will be no impact on the published library when changing this function definition.